libtiledbvcf/src/stats/variant_stats_reader.cc

```diff
@@ -45,13 +46,230 @@ inline int pos_comparator(const void* a, const void* b) {
+  // allele count.
+  auto next_allele = pos_map.second.find(allele);
+
+  // First multiply by 1.0 to force a float type, then look up AC from the
+  // above iterator. Substitute 0 for the AC value if it is absent in pos_map.
+  // Divide by AN (pos_map.first) to get AF.
+  if (next_allele == pos_map.second.end()) {
+    return {0, 0, pos_map.first};
+  }
+  return {
+      1.0 * next_allele->second / pos_map.first,
+      next_allele->second,
+      pos_map.first};
+}
+
+std::tuple<float, uint32_t, uint32_t> AFMap::af(
+    uint32_t pos, const std::string& allele, size_t num_samples) {
+  // calculate af = ac / an
+  std::unordered_map<
+      uint32_t,
+      std::pair<int, std::unordered_map<std::string, int>>>::const_iterator
+      pos_map_iterator = ac_map_.find(pos);
+
+  // Return -1.0 if the allele was not called
+  if (pos_map_iterator == ac_map_.end()) {
+    return {-1.0, 0, 0};
+  }
+
+  const std::pair<int, std::unordered_map<std::string, int>>& pos_map =
+      pos_map_iterator->second;
+
+  // We don't know that allele was called in this sample. Ask nicely for the
+  // allele count.
+  decltype(pos_map.second)::const_iterator next_allele =
+      pos_map.second.find(allele);
+
+  // First multiply by 1.0 to force a float type, then look up AC from the
+  // above iterator. Substitute 0 for the AC value if it is absent in pos_map.
+  // Divide by AN (pos_map.first) to get AF.
+  if (next_allele == pos_map.second.end()) {
+    return {0, 0, num_samples * 2};
+  }
+  return {
+      1.0 * next_allele->second / num_samples / 2,
+      next_allele->second,
+      num_samples * 2};
+}
+
+std::tuple<float, uint32_t, uint32_t> AFMap::af_v3(
+    uint32_t pos, const std::string& allele) {
+  // calculate af = ac / an
+  std::unordered_map<
+      uint32_t,
+      std::pair<int, std::unordered_map<std::string, int>>>::const_iterator
+      pos_map_iterator = ac_map_.find(pos);
+
+  // Return -1.0 if the allele was not called
+  if (pos_map_iterator == ac_map_.end()) {
+    return {-1.0, 0, 0};
+  }
+
+  const auto pos_map = pos_map_iterator->second;
+
+  // We don't know that allele was called in this sample. Ask nicely for the
+  // allele count.
+  auto next_allele = pos_map.second.find(allele);
+  bool is_ref = allele == "ref";
+
+  // First multiply by 1.0 to force a float type, then look up AC from the
+  // above iterator. Substitute 0 for the AC value if it is absent in pos_map.
+  // Divide by AN (pos_map.first) to get AF.
+  if (next_allele == pos_map.second.end()) {
+    return {
+        is_ref ? 1.0 * ac_sum_ / (pos_map.first + an_sum_) : 0,
+        is_ref ? ac_sum_ : 0,
+        pos_map.first + an_sum_};
+  }
+  return {
+      1.0 * (next_allele->second + (is_ref ? ac_sum_ : 0)) /
+          (pos_map.first + an_sum_),
+      next_allele->second + (is_ref ? ac_sum_ : 0),
+      (pos_map.first + an_sum_)};
+}
+
+std::tuple<float, uint32_t, uint32_t> AFMap::af_v3(
```

Comment:
Same comment as above for combining versions of `af_v3`.

Reply:
same as above: we wouldn't be making it much less succinct, and would be adding runtime complexity by combining them